server/output/filesystem/filesystem.go

```diff
@@ -0,0 +1,129 @@
+	err := createDirectory(directory)
+	if err != nil {
+		return FilesystemOutput{}, fmt.Errorf("Failed to create directory %s for filesystem output: %w", directory, err)
+	}
+
+	filesystemOutput := FilesystemOutput{
+		directory:             directory,
+		metaFiles:             make(map[uuid.UUID]metaFile),
+		maxFileAge:            maxFileAge,
+		maxFileSize:           maxFileSize,
+		metaFileCreationMutex: &sync.Mutex{},
+	}
+
+	return filesystemOutput, nil
+}
+
+// FilesystemOutput is an event output that writes events to the filesystem
+type FilesystemOutput struct {
+	directory             string
+	metaFiles             map[uuid.UUID]metaFile
+	maxFileAge            time.Duration
+	maxFileSize           int
+	metaFileCreationMutex *sync.Mutex
+}
+
+// Close closes are underlying file descriptors for the FilesystemOutput
+// Close should only be called once after Consume is guaranteed not to be called again
+func (filesystemOutput FilesystemOutput) Close() {
+
+	metaFiles := filesystemOutput.metaFiles
+	filesystemOutput.metaFiles = make(map[uuid.UUID]metaFile)
+
+	for _, metaFile := range metaFiles {
+		metaFile.Close()
+	}
+}
+
+// Consume takes the events and writes them to a channel for processing
+func (filesystemOutput FilesystemOutput) Consume(events []events.Event) error {
+
+	// All events in a single call will have the same group and sensor
+	group := events[0].GetGroup()
+
+	metaFile, err := filesystemOutput.getGroupMetaFile(group)
+	if err != nil {
+		return fmt.Errorf("Failed to write event to filesystem: %w", err)
+	}
+
+	return metaFile.Consume(events)
+}
+
+func (filesystemOutput FilesystemOutput) getGroupMetaFile(group uuid.UUID) (metaFile, error) {
+	var metaFileInstance metaFile
+	var ok bool
+	if metaFileInstance, ok = filesystemOutput.metaFiles[group]; !ok {
+		return filesystemOutput.createAndStoreGroupMetaFile(group)
+	}
+
+	return metaFileInstance, nil
+}
+
+func (filesystemOutput FilesystemOutput) createAndStoreGroupMetaFile(group uuid.UUID) (metaFile, error) {
+	filesystemOutput.metaFileCreationMutex.Lock()
+	defer filesystemOutput.metaFileCreationMutex.Unlock()
+
+	var metaFileInstance metaFile
+	var ok bool
+	if metaFileInstance, ok = filesystemOutput.metaFiles[group]; !ok {
+
+		metaFileDirectoryPath := filepath.Join(filesystemOutput.directory, group.String())
+		err := createDirectory(metaFileDirectoryPath)
+		if err != nil {
+			return metaFileInstance, fmt.Errorf("Failed to create directory %s for filesystem output: %w", metaFileDirectoryPath, err)
+		}
+
+		metaFileInstance, err = newMetaFile(metaFileDirectoryPath, filesystemOutput.maxFileAge, filesystemOutput.maxFileSize)
+		if err != nil {
+			return metaFileInstance, fmt.Errorf("Failed to create file for writing: %w", err)
+		}
+
+		filesystemOutput.metaFiles[group] = metaFileInstance
+	}
+
+	return metaFileInstance, nil
+}
+
+func createDirectory(path string) error {
+	_, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		err = os.Mkdir(path, 0755)
```

Comment:
Files are created without permissions to other groups. Shouldn't give them for directories either to stay consistent with least privilege.